File: XstarS.GuidGenLib.UnitTests/TestHelpers/DisposeAction.cs
using System;

namespace XNetEx;

internal sealed class DisposeAction(Action<bool> action)
    : IDisposable
{
    private readonly Action<bool> Action = action ??
        throw new ArgumentNullException(nameof(action));

    private bool IsDisposed = false;

    ~DisposeAction()
    {
        this.Dispose(disposing: false);
    }

    public void Dispose()
    {
        this.Dispose(disposing: true);
        GC.SuppressFinalize(this);
    }

    private void Dispose(bool disposing)
    {
        if (!this.IsDisposed)
        {
            this.Action.Invoke(disposing);
            this.IsDisposed = true;
        }
    }
}

File: XstarS.GuidModule.UnitTests/TestHelpers/DisposeAction.fs
namespace XNetEx.FSharp.Control

open System

[<Sealed>]
type internal DisposeAction(action: bool -> unit) =
    inherit obj()

    let mutable isDisposed = false

    override this.Finalize() =
        try
            this.Dispose(false)
        finally
            base.Finalize()

    member this.Dispose() =
        this.Dispose(true)
        GC.SuppressFinalize(this)

    member private _.Dispose(disposing: bool) =
        if not isDisposed then
            action disposing
            isDisposed <- true

    interface IDisposable with
        member this.Dispose() =
            this.Dispose()
